sklearn/tree/_criterion.pyx

```diff
@@ -962,6 +965,289 @@ cdef class MSE(RegressionCriterion):
+        self.node_medians = NULL
+
+        # Allocate memory for the accumulators
+        self.node_medians = <double*> calloc(n_outputs, sizeof(double))
+
+        if (self.node_medians == NULL):
+            raise MemoryError()
+
+        self.left_child_heaps = np.empty(n_outputs, dtype='object')
+        self.right_child_heaps = np.empty(n_outputs, dtype='object')
+
+    cdef void init(self, DOUBLE_t* y, SIZE_t y_stride, DOUBLE_t* sample_weight,
+                   double weighted_n_samples, SIZE_t* samples, SIZE_t start,
+                   SIZE_t end) nogil:
+        """Initialize the criterion at node samples[start:end] and
+           children samples[start:start] and samples[start:end]."""
+
+        # Initialize fields
+        self.y = y
+        self.y_stride = y_stride
+        self.sample_weight = sample_weight
+        self.samples = samples
+        self.start = start
+        self.end = end
+        self.n_node_samples = end - start
+        self.weighted_n_samples = weighted_n_samples
+        self.weighted_n_node_samples = 0.
+
+        cdef SIZE_t i
+        cdef SIZE_t p
+        cdef SIZE_t k
+        cdef DOUBLE_t y_ik
+        cdef DOUBLE_t w = 1.0
+        # Fill accumulators with MedianHeaps
+        with gil:
+            for k in range(self.n_outputs):
+                self.left_child_heaps[k] = WeightedMedianHeap(self.n_node_samples)
+                self.right_child_heaps[k] = WeightedMedianHeap(self.n_node_samples)
+
+        cdef void** left_child_heaps = <void**> self.left_child_heaps.data
+        cdef void** right_child_heaps = <void**> self.right_child_heaps.data
+
+        for p in range(start, end):
+            i = samples[p]
+
+            if sample_weight != NULL:
+                w = sample_weight[i]
+
+            for k in range(self.n_outputs):
+                y_ik = y[i * y_stride + k]
+
+                # push all values to the right side,
+                # since pos = start initially anyway
+                (<WeightedMedianHeap> right_child_heaps[k]).push(y_ik, w)
+
+            self.weighted_n_node_samples += w
+
+        # calculate the node medians
+        for k in range(self.n_outputs):
+            (<WeightedMedianHeap> right_child_heaps[k]).get_median(&(self.node_medians[k]))
+
+        # Reset to pos=start
+        self.reset()
+
+    cdef void reset(self) nogil:
+        """Reset the criterion at pos=start."""
+
+        cdef SIZE_t i
+        cdef SIZE_t k
+        cdef DOUBLE_t popped_value
+        cdef DOUBLE_t popped_weight
+
+        cdef void** left_child_heaps = <void**> self.left_child_heaps.data
+        cdef void** right_child_heaps = <void**> self.right_child_heaps.data
+
+        self.weighted_n_left = 0.0
+        self.weighted_n_right = self.weighted_n_node_samples
+        self.pos = self.start
+
+        # reset the medianheaps, left should have no elements and
+        # right should have all elements.
+
+        for k in range(self.n_outputs):
```

Comment:
You should have a reset method on the median heap that you call, which internally will clear everything. It will clarify what is happening here.